subprojects/gradle/build-on-target/src/main/kotlin/com/avito/buildontarget/BuildOnTargetPlugin.kt

```diff
@@ -0,0 +1,91 @@
+package com.avito.buildontarget
+
+import com.android.build.gradle.api.ApplicationVariant
+import com.android.build.gradle.api.TestVariant
+import com.avito.android.getApkFile
+import com.avito.android.withAndroidApp
+import com.avito.git.Branch
+import com.avito.git.gitState
+import com.avito.kotlin.dsl.getBooleanProperty
+import com.avito.kotlin.dsl.getMandatoryIntProperty
+import com.avito.kotlin.dsl.getMandatoryStringProperty
+import com.avito.utils.gradle.envArgs
+import com.avito.utils.logging.ciLogger
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.file.RegularFile
+import org.gradle.api.internal.provider.Providers
+import org.gradle.api.provider.Provider
+import org.gradle.kotlin.dsl.register
+import java.io.File
+
+class BuildOnTargetPlugin : Plugin<Project> {
+
+    private val ciTaskGroup = "ci"
+
+    override fun apply(project: Project) {
+
+        val logger = project.ciLogger
+        val env = project.envArgs
+        val gitState = project.gitState { logger.info(it) }
+
+        @Suppress("UnstableApiUsage")
+        val targetBranch: Provider<Branch> = gitState.flatMap { state ->
+            val targetBranch = state.targetBranch
+            if (targetBranch != null) {
+                Providers.of(targetBranch)
+            } else {
+                Providers.notDefined()
+            }
+        }
+
+        project.withAndroidApp { appExtension ->
+            appExtension.testVariants.all { testVariant: TestVariant ->
+                val testedVariant: ApplicationVariant =
+                    testVariant.testedVariant as ApplicationVariant
+
+                val buildOnTargetCommitTask =
+                    project.tasks.register<BuildOnTargetCommitForTestTask>(buildOnTargetTaskName()) {
```

Comment:
Do we register the same task for all tested variants? Could we get a conflict because of `taskName`?

Reply:
nope, it's a single task for all configurations; it's on purpose